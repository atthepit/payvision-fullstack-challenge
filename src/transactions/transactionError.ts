class TransactionError extends Error {}

export default TransactionError;
