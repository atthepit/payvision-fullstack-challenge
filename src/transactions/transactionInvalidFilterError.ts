import TransactionError from "./transactionError";

class TransactionInvalidFilterError extends TransactionError {}

export default TransactionInvalidFilterError;
