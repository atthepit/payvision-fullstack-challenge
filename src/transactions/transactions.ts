import axios, { AxiosRequestConfig } from "axios";
import TransactionInvalidFilterError from "./transactionInvalidFilterError";
import TransactionError from "./transactionError";

if (!process.env.API_USERNAME || !process.env.API_PASSWORD) {
  throw new Error("Missing API credentials");
}

if (!process.env.API_BASE_URL) {
  throw new Error("Missing API configuration (API_BASE_URL)");
}

const CREDENTIALS = btoa(
  `${process.env.API_USERNAME}:${process.env.API_PASSWORD}`
);

const BASE_URL = process.env.API_BASE_URL;

interface Filters {
  action?: string;
  currencyCode?: string;
  orderBy?: string;
  [propName: string]: any;
}

function formatURL(filters: Filters) {
  let connector = "?";
  return Object.keys(filters).reduce((url, key) => {
    url = `${url}${connector}${key}=${filters[key]}`;
    connector = "&";
    return url;
  }, BASE_URL);
}

export interface Transaction {
  action: string;
  amount: number;
  brandId: number;
  card: {
    expiryMonth: string;
    expiryYear: string;
    firstSixDigits: string;
    lastFourDigits: string;
    holderName: string;
  };
  currencyCode: string;
  id: string;
  trackingCode: string;
}

export function search(filters?: Filters): Promise<Transaction[]> {
  filters = filters || {};
  const url = formatURL(filters);
  const config = {
    headers: {
      Authorization: `Basic ${CREDENTIALS}`
    }
  };

  return tryFetchTransaction(url, config);
}

async function tryFetchTransaction(url: string, config: AxiosRequestConfig) {
  try {
    return await fetchTransactions(url, config);
  } catch (error) {
    const { message } = error.response.data;
    if (message === "INVALID_FILTER") {
      throw new TransactionInvalidFilterError("One or more invalid filters");
    } else {
      throw new TransactionError(`Unexpected error: ${message}`);
    }
  }
}

async function fetchTransactions(url: string, config: AxiosRequestConfig) {
  const response = await axios.get(url, config);
  const transactions = response.data as Transaction[];
  return transactions;
}
