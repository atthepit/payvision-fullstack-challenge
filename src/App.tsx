import React from "react";
import { search, Transaction, Filters } from "./transactions/transactions";
import Search from "./Search";

interface Props {}

interface State {
  transactions: Transaction[];
}

class App extends React.Component<Props, State> {
  public state = {
    transactions: [] as Transaction[]
  };

  public componentDidMount() {
    search().then(transactions => {
      this.setState({ transactions });
    });
  }

  public handleSearch = (filters: Filters) => {
    search(filters).then(transactions => {
      this.setState({ transactions });
    });
  };

  public render() {
    return (
      <div>
        <Search onSearch={this.handleSearch} />
        <p data-testid="transactions">
          You have {this.state.transactions.length} transactions!
        </p>
      </div>
    );
  }
}

export default App;
