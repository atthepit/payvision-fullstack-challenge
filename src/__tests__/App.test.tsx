import React from "react";
import {
  render,
  fireEvent,
  cleanup,
  waitForElement
} from "react-testing-library";
import App from "../App";
import { search } from "../transactions/transactions";
jest.mock("../transactions/transactions");

afterEach(cleanup);
afterEach(() => {
  search.mockReset();
});

test("handleSearch", async () => {
  search.mockImplementation(() => Promise.resolve([]));
  const { getByLabelText, getByText, getByTestId } = render(<App />);
  let resultNode = await waitForElement(() => getByTestId("transactions"));
  expect(resultNode.textContent).toMatch("0");

  const actionNode = getByLabelText("Action") as HTMLSelectElement;
  const currencyCodeNode = getByLabelText("Currency Code") as HTMLSelectElement;
  const searchButton = getByText("Search");

  let action = "payment";
  let currencyCode = "EUR";
  actionNode.value = action;
  currencyCodeNode.value = currencyCode;

  search.mockImplementationOnce(() => Promise.resolve([{}]));
  fireEvent.click(searchButton);
  resultNode = await waitForElement(() => getByTestId("transactions"));
  expect(resultNode.textContent).toMatch("1");
  expect(search).toHaveBeenLastCalledWith({ action, currencyCode });

  action = "payment";
  currencyCode = "USD";
  actionNode.value = action;
  currencyCodeNode.value = currencyCode;

  search.mockImplementationOnce(() => Promise.resolve([{}, {}]));
  fireEvent.click(searchButton);
  resultNode = await waitForElement(() => getByTestId("transactions"));
  expect(resultNode.textContent).toMatch("2");
  expect(search).toHaveBeenLastCalledWith({ action, currencyCode });

  expect(search).toHaveBeenCalledTimes(3);
});
