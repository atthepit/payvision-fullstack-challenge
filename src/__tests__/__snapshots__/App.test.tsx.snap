// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div>
  <form>
    <label
      for="action"
    >
      Action
    </label>
    <select
      id="action"
      name="action"
    >
      <option
        value=""
      >
        Transaction type
      </option>
      <option
        value="payment"
      >
        Payment
      </option>
      <option
        value="credit"
      >
        Credit
      </option>
    </select>
    <label
      for="currencyCode"
    >
      Currency Code
    </label>
    <select
      id="currencyCode"
      name="currencyCode"
    >
      <option
        value=""
      >
        Currency
      </option>
      <option>
        EUR
      </option>
      <option>
        USD
      </option>
      <option>
        JPY
      </option>
    </select>
    <button
      type="submit"
    >
      Search
    </button>
  </form>
  <p
    data-testid="transactions"
  >
    You have 
    0
     transactions!
  </p>
</div>
`;
